src/Common/src/System/Collections/Generic/BidirectionalDictionary.cs

```diff
@@ -0,0 +1,50 @@
+// Licensed to the .NET Foundation under item1 or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+
+namespace System.Collections.Generic
+{
+    internal sealed class BidirectionalDictionary<T1, T2> : IEnumerable<Tuple<T1, T2>>
+    {
+        private readonly Dictionary<T1, T2> _forward;
+        private readonly Dictionary<T2, T1> _backward;
+
+        public BidirectionalDictionary(int capacity)
+        {
+            _forward = new Dictionary<T1, T2>(capacity);
+            _backward = new Dictionary<T2, T1>(capacity);
+        }
+
+        public void Add(T1 item1, T2 item2)
+        {
+            Debug.Assert(!_backward.ContainsKey(item2), " No added item1 should ever have existing item2");
+            _forward.Add(item1, item2);
+            _backward.Add(item2, item1);
+        }
+
+        public bool TryGetForward(T1 item1, out T2 item2)
```

Comment:
If T1 != T2 these would be unambiguous overloads of just TryGet(T1 item1, out T2 item2), TryGet(T2 item2, out T1 item1).  And I vaguely feel like a != constraint can be applied (or maybe I merely recall wishing it could be done).

It's just a bit gnarly to have to remember which way you decided was "forward".  Bidi is bidi, you're not supposed to care :)